// jira.js/out/version2/issueSecuritySchemes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IssueSecuritySchemes = void 0;
const tslib_1 = require("tslib");
class IssueSecuritySchemes {
    constructor(client) {
        this.client = client;
    }
    getIssueSecuritySchemes(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/issuesecurityschemes',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getIssueSecurityScheme(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/2/issuesecurityschemes/${id}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.IssueSecuritySchemes = IssueSecuritySchemes;
